lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserSwiftDescriptorFinder.cpp

```diff
@@ -0,0 +1,225 @@
+        m_type_system(type_system), m_mangled_name(mangled_name),
+        m_die_ref(die_ref) {}
+
+  ~DWARFFieldDescriptorImpl() override = default;
+
+  // TODO: implement this.
+  swift::Demangle::NodePointer demangleSuperclass() override { return nullptr; }
+
+  std::vector<std::unique_ptr<swift::reflection::FieldRecordBase>>
+  getFieldRecords() override {
+    if (!Target::GetGlobalProperties().GetEnableFullDwarfDebugging())
+      return {};
+    auto *dwarf =
+        llvm::dyn_cast<SymbolFileDWARF>(m_type_system.GetSymbolFile());
+    auto *dwarf_parser = m_type_system.GetDWARFParser();
+    if (!dwarf || !dwarf_parser)
+      return {};
+
+    auto die = dwarf->GetDIE(m_die_ref);
+    if (!die)
+      return {};
+
+    // For now we only support structs.
+    if (Kind != swift::reflection::FieldDescriptorKind::Struct)
+      return {};
+
+    std::vector<std::unique_ptr<swift::reflection::FieldRecordBase>> fields;
+    for (DWARFDIE child_die : die.children()) {
+      auto tag = child_die.Tag();
+      if (tag != DW_TAG_member)
+        continue;
+      const auto *member_field_name =
+          child_die.GetAttributeValueAsString(llvm::dwarf::DW_AT_name, "");
+      auto *member_type = dwarf_parser->GetTypeForDIE(child_die);
+      auto member_mangled_typename =
+          member_type->GetForwardCompilerType().GetMangledTypeName();
+
+      // Only matters for enums, so set to false for structs.
+      bool is_indirect_case = false;
+      // Unused by type info construction.
+      bool is_var = false;
+      fields.emplace_back(std::make_unique<DWARFFieldRecordImpl>(
+          is_indirect_case, is_var, ConstString(member_field_name),
+          member_mangled_typename));
+    }
+    return fields;
+  }
+};
+} // namespace
+
+/// Constructs a builtin type descriptor from DWARF information.
+std::unique_ptr<swift::reflection::BuiltinTypeDescriptorBase>
+DWARFASTParserSwift::getBuiltinTypeDescriptor(
+    const swift::reflection::TypeRef *TR) {
+  if (!Target::GetGlobalProperties().GetEnableFullDwarfDebugging())
+    return nullptr;
+
+  auto pair = getTypeAndDie(m_swift_typesystem, TR);
+  if (!pair)
+    return nullptr;
+  auto &[type, die] = *pair;
+
+  if (die.Tag() == llvm::dwarf::DW_TAG_structure_type) {
+    auto child = die.GetFirstChild();
+    if (child.Tag() != llvm::dwarf::DW_TAG_variant_part)
+      return nullptr;
+  } else if (die.Tag() != llvm::dwarf::DW_TAG_base_type)
+    return nullptr;
+
+  auto byte_size =
+      die.GetAttributeValueAsUnsigned(DW_AT_byte_size, LLDB_INVALID_ADDRESS);
+  if (byte_size == LLDB_INVALID_ADDRESS)
+    return {};
+  auto alignment = die.GetAttributeValueAsUnsigned(
+      DW_AT_alignment, byte_size == 0 ? 1 : byte_size);
+
+  // TODO: this seems simple to calculate but maybe we should encode the stride
+  // in DWARF? That's what reflection metadata does.
+  unsigned stride = ((byte_size + alignment - 1) & ~(alignment - 1));
```

Comment:
Should this use llvm::alignTo for better readability? https://llvm.org/doxygen/namespacellvm.html#a0654c8f28b6096222fdc77c843576ad6

Reply:
I actually tried using that before, but it isn't the same operation as `llvm::alignTo` (or any other function in math extras that I could find).